partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rel/RelMaterialized.kt

```diff
@@ -0,0 +1,48 @@
+package org.partiql.eval.internal.operator.rel
+
+import org.partiql.eval.internal.Environment
+import org.partiql.eval.internal.Record
+import org.partiql.eval.internal.operator.Operator
+
+/**
+ * For [Operator.Relation]'s that MUST materialize data in order to execute [hasNext], this abstract class caches the
+ * result of [materializeNext] to implement both [hasNext] and [next].
+ */
+internal abstract class RelMaterialized : Operator.Relation {
```

Comment:
Was looking at [Guava PeekingIterator](https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/Iterators.java#L1185-L1211) and [Apache PeekingIterator](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/iterators/PeekingIterator.html). 

This is interesting because it's an abstract class (rather than a wrapper over the input relation). I _think_ the abstract class fits our use case better.

We could call this RelPeeking and `materializeNext() -> peek()` then our implementations line up with the Guava and Apache impls well.

I don't think we need a _nextIsReady because in kotlin we can have a nullable Record (Record?) and whether or not we set this thing to null tells us if something is next/nextIsReady.

```kotlin
override fun hasNext(): Boolean {
    if (_next != null) return true
    _next = peek()
    return _next != null
}

override fun next(): Record {
    val next = _next ?: error("...")
    _next = null
    return next
}
```